File: apps/windows/TeslaSync.App.Core/Auth/OidcConfig.cs
using System.Collections.ObjectModel;

namespace TeslaSync.App.Core.Auth;

/// <summary>
/// Immutable OIDC public-client configuration for the Authorization-Code-with-PKCE
/// flow against Authentik (ADR-008, authentik-native-clients runbook). A native app
/// is a <b>public</b> client — there is intentionally no client secret.
///
/// Endpoints are supplied explicitly (they come from the provider's discovery
/// document); this type does not guess them from an issuer so a misconfigured tenant
/// fails loudly rather than silently hitting the wrong URL.
/// </summary>
public sealed class OidcConfig
{
    /// <summary>openid + identity scopes, plus offline_access for a refresh token.</summary>
    public static IReadOnlyList<string> DefaultScopes { get; } =
        new ReadOnlyCollection<string>(["openid", "profile", "email", "offline_access"]);

    /// <summary>Creates a config; throws when any required endpoint/value is missing.</summary>
    public OidcConfig(
        string clientId,
        string redirectUri,
        string authorizationEndpoint,
        string tokenEndpoint,
        string? revocationEndpoint = null,
        IReadOnlyList<string>? scopes = null)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(clientId);
        ArgumentException.ThrowIfNullOrWhiteSpace(redirectUri);
        ArgumentException.ThrowIfNullOrWhiteSpace(authorizationEndpoint);
        ArgumentException.ThrowIfNullOrWhiteSpace(tokenEndpoint);

        ClientId = clientId;
        RedirectUri = redirectUri;
        AuthorizationEndpoint = authorizationEndpoint;
        TokenEndpoint = tokenEndpoint;
        RevocationEndpoint = revocationEndpoint;
        Scopes = scopes is { Count: > 0 } ? scopes : DefaultScopes;
    }

    /// <summary>The per-platform public client id (e.g. <c>teslasync-windows</c>).</summary>
    public string ClientId { get; }

    /// <summary>The exact, pre-registered redirect URI the authorization response returns to.</summary>
    public string RedirectUri { get; }

    /// <summary>Absolute URL of the provider authorize endpoint.</summary>
    public string AuthorizationEndpoint { get; }

    /// <summary>Absolute URL of the provider token endpoint.</summary>
    public string TokenEndpoint { get; }

    /// <summary>Absolute URL of the provider revocation endpoint, or <see langword="null"/>.</summary>
    public string? RevocationEndpoint { get; }

    /// <summary>Requested scopes; <c>offline_access</c> is required to obtain a refresh token.</summary>
    public IReadOnlyList<string> Scopes { get; }
}

File: apps/shared/core/src/commonMain/kotlin/io/teslasync/shared/core/auth/OidcConfig.kt
package io.teslasync.shared.core.auth

/**
 * Immutable OIDC public-client configuration for the Authorization-Code-with-PKCE
 * flow against Authentik (ADR-008, P0/0009 runbook). A native app is a **public**
 * client — there is intentionally no client secret.
 *
 * Endpoints are supplied explicitly (they come from the provider's discovery
 * document); this type does not guess them from an issuer so a misconfigured tenant
 * fails loudly rather than silently hitting the wrong URL.
 *
 * @property clientId the per-platform public client id (e.g. `teslasync-android`).
 * @property redirectUri the exact, pre-registered redirect URI (App/Universal Link
 *   or custom scheme) the authorization response returns to.
 * @property authorizationEndpoint absolute URL of the provider authorize endpoint.
 * @property tokenEndpoint absolute URL of the provider token endpoint.
 * @property revocationEndpoint absolute URL of the provider revocation endpoint, or
 *   `null` when the tenant does not expose one (sign-out then only clears storage).
 * @property scopes requested scopes; `offline_access` is required to obtain a
 *   refresh token.
 */
public class OidcConfig(
    public val clientId: String,
    public val redirectUri: String,
    public val authorizationEndpoint: String,
    public val tokenEndpoint: String,
    public val revocationEndpoint: String? = null,
    public val scopes: List<String> = DEFAULT_SCOPES,
) {
    public companion object {
        /** openid + identity scopes, plus offline_access for a refresh token. */
        public val DEFAULT_SCOPES: List<String> =
            listOf("openid", "profile", "email", "offline_access")
    }
}
